packages/dds/tree/src/simple-tree/schemaFactoryRecursive.ts

```diff
@@ -3,21 +3,193 @@
  * Licensed under the MIT License.
  */
 
-import { FlexTreeNode, isFlexTreeNode } from "../feature-libraries/index.js";
+import {
+	FlexListToUnion,
+	FlexTreeNode,
+	Unenforced,
+	isFlexTreeNode,
+} from "../feature-libraries/index.js";
 import { RestrictiveReadonlyRecord } from "../util/index.js";
 import {
+	AllowedTypes,
+	ApplyKind,
+	FieldKind,
+	FieldSchema,
 	ImplicitAllowedTypes,
 	ImplicitFieldSchema,
-	InsertableObjectFromSchemaRecord,
 	InsertableTreeNodeFromImplicitAllowedTypes,
+	NodeFromSchema,
 	NodeKind,
-	ObjectFromSchemaRecord,
-	TreeMapNode,
+	TreeNodeFromImplicitAllowedTypes,
+	TreeNodeSchema,
 	TreeNodeSchemaClass,
 	WithType,
 } from "./schemaTypes.js";
 import { SchemaFactory, type ScopedSchemaName } from "./schemaFactory.js";
 import { TreeArrayNode } from "./treeArrayNode.js";
+import { TreeArrayNodeBase, TreeNode, Unhydrated } from "./types.js";
+
+/*
+ * TODO:
+ * Below are a bunch of "unsafe" versions of types from "schemaTypes.ts".
+ * These types duplicate the ones in "schemaTypes.ts", except with some of the extends clauses unenforced.
+ * This is not great for type safety or maintainability.
+ * Eventually it would be great to do at least one of the following:
+ * 1. Find a way to avoid needing these entirely, possibly by improving TSC's recursive type support.
+ * 2. Deduplicate the safe and unsafe types (possibly by having the safe one call the unsafe ones, or some other trick).
```

Comment:
are there known obstacles to 2? Or is it more along the lines of (probably justified) paranoia that it will change typescript behavior in some way? I suppose it does mean that the unsafe type variants get exposed more than we'd like at this point. That's a good enough reason for me to be ok with what's implemented.

Mostly just wondering why you haven't opted to do that in this PR since it seems like a straightforward change.

Reply:
Doing 2 should involve a lot of careful testing, evaluation of IntelliSense, stabilizing the unsafe APIs as public (they aren't yet), deciding that its actually more maintainable etc. I'm not convinced and of the obious ways to do it are a win, or the best way to go forward. Ex: maybe just removing the type constraints in a few cases and merging those makes more sense, or maybe fixing TSC is the best approach. For now though, the different release stability levels blocks a lot of the approaches, so I figured punting it would be fine.

Reply:
Yup, seems totally reasonable to me.